compiler/src/dotty/tools/dotc/transform/GenericSignatures.scala

```diff
@@ -475,4 +453,28 @@ object GenericSignatures {
         }
       else x
   }
+
+  private def collectMethodParams(mtd: MethodOrPoly)(using Context): (List[TypeParamInfo], List[Type], Type) = 
+    val tparams = ListBuffer.empty[TypeParamInfo]
+    val vparams = ListBuffer.empty[Type]
+
+    @tailrec def recur(mtd: MethodOrPoly): Type = mtd match
```

Comment:
If you accept any `Type` here, you can factor out the ending cases in a much more streamlined way. Something like:
```scala
    @tailrec def recur(tpe: Type): Type = tpe match
      case tpe: MethodType =>
        vparams ++= ...
        recur(tpe.resType)
      case tpe: PolyType =>
        ...
        recur(tpe.resType)
      case _ =>
        tpe
```